Support W4A8 quantization for vllm

HandH1998 opened the pull request:
We have proposed a W4A8 quantization solution QQQ and integrated it into vllm.  QQQ can not only achieve the similar performance of the leading W4A8, W8A8, and W4A16 quantization methods but also significantly accelerate inference—achieving up to 2.24x, 2.10x, and 1.25x speed boosting compared to FP16, W8A8, and W4A16(Marlin), respectively. 

## News or Update
- [2024/06/17] **Update!!!** We release the QQQ [paper](https://arxiv.org/pdf/2406.09904) on arXiv.
- [2024/06/07] We fuse dynamic activation quantization into `QQQLinearMethod` to support QQQ for all the models. Note that this will reduce some inference performance. You should use this [repo](https://github.com/HandH1998/vllm/tree/w4a8-fusion) to reproduce the inference speed results of our paper.
- [2024/06/03] We integrate QQQ into vLLM and release the code.

## Usage
You can export the quantized model weights with this repo [QQQ](https://github.com/HandH1998/QQQ)(only support llama for now). Our paper will be published on arXiv soon. Here is an offline inference example.
```
from vllm import LLM, SamplingParams

# Sample prompts.
prompts = [
    "Hello, my name is",
    "The president of the United States is",
    "The capital of France is a",
    "A pig",
]
# Create a sampling params object.
sampling_params = SamplingParams(temperature=0.8, top_p=0.95)
model = your_quantized_model_path
tokenizer = your_tokenizer_path

# Create an LLM.
llm = LLM(
    model=model,
    tokenizer=tokenizer,
)
# Generate texts from the prompts. The output is a list of RequestOutput objects
# that contain the prompt, generated text, and other information.
outputs = llm.generate(prompts, sampling_params)
# Print the outputs.
for output in outputs:
    prompt = output.prompt
    generated_text = output.outputs[0].text
    print(f"Prompt: {prompt!r}, Generated text: {generated_text!r}")

  ```
 ## Experiments
 Here we provide our experiment results using this [repo](https://github.com/HandH1998/vllm/tree/w4a8-fusion).
 ### Settings
 - vllm v0.4.1
 - 1 A100 80G
 - CUDA 11.8
 ### Model Performance
 We evaluated the model performance on WikiText2 and five zero-shot tasks.
<img width="738" alt="model_performace" src="https://github.com/vllm-project/vllm/assets/44199326/0b532183-9f78-4651-8d34-b01ccfcd5aa9">
 ### Throughput
We conducted the same-batch throughput comparison of quantized LLaMA-2 models under various batch sizes. The input sequence length is 1024 and the output sequence length is 128.
<img width="1694" alt="speedup" src="https://github.com/user-attachments/assets/26900e16-21a3-4d4c-8ce3-ed9830beb8a3">

### W4A8 GEMM Performance
We implement the W4A8 GEMM based on [Marlin GEMM](https://github.com/IST-DASLab/marlin). Thanks to their great work! Here is the speedup over PyTorch FP16 GEMM (Calling CUTLASS) of all GEMMs under different numbers of input tokens. The weight matrix size is (N=8192, K=21760).
<img width="473" alt="gemm_performance" src="https://github.com/vllm-project/vllm/assets/44199326/fe8cc83b-4a77-4791-a0f9-c5ac717d174c">

robertgshaw2-neuralmagic commented:
Thanks for the PR!

The team for NM will review this. We are going to have to split this work out into a few pieces.

Will get back to you

zhyncs commented:
> remove the A8 compute from `LayerNorm`, `Attention`, and `SiluAndMul`

@robertgshaw2-neuralmagic This will reduce some performance gains. Could you share a more general approach and timeline? Perhaps our team can collaborate to implement the corresponding features in this PR, and then you can modify it into a general version later. The existing W4A8 in the industry does not meet our accuracy requirements. That's why we do this. In order to maximize performance while ensuring accuracy, all efforts are made to make this quantization truly usable in an online production environment, not just for publishing papers.

mgoin commented:
@zhyncs The goal of making this W4A8 optimization "production-ready" is exactly why I also think it is a good idea to land the first step as simply having this only be enabled for Linear modules as `QQQLinearMethod`. With this, we can enable the method for all models, rather than just supporting Llama.

We have an RFC tracking Int8 W8A8 that lays out some of the high-level goals, including graph fusion through torch.compile https://github.com/vllm-project/vllm/issues/3975. We have been landing many kernels recently and hope to share the torch.compile prototype soon!

HandH1998 commented:
> @zhyncs The goal of making this W4A8 optimization "production-ready" is exactly why I also think it is a good idea to land the first step as simply having this only be enabled for Linear modules as `QQQLinearMethod`. With this, we can enable the method for all models, rather than just supporting Llama.
> 
> We have an RFC tracking Int8 W8A8 that lays out some of the high-level goals, including graph fusion through torch.compile #3975. We have been landing many kernels recently and hope to share the torch.compile prototype soon!

I see what you want to do. Hope you can finish supporting dynamic Activation Quantization successfully. I will try to split the activation quantization with `LayerNorm`, `Attention`, and `SiluAndMul`, and put it in `QQQLinearMethod`. Note that this will reduce some inference performance.

robertgshaw2-neuralmagic commented:
> > @zhyncs The goal of making this W4A8 optimization "production-ready" is exactly why I also think it is a good idea to land the first step as simply having this only be enabled for Linear modules as `QQQLinearMethod`. With this, we can enable the method for all models, rather than just supporting Llama.
> > We have an RFC tracking Int8 W8A8 that lays out some of the high-level goals, including graph fusion through torch.compile #3975. We have been landing many kernels recently and hope to share the torch.compile prototype soon!
> 
> I see what you want to do. Hope you can finish supporting dynamic Activation Quantization successfully. I will try to split the activation quantization with `LayerNorm`, `Attention`, and `SiluAndMul`, and put it in `QQQLinearMethod`. Note that this will reduce some inference performance.

SG - then we can add in the quantized execution of non-Linear layers as part of our broader project

HandH1998 commented:
@robertgshaw2-neuralmagic We have finished the mentioned work. This may help you achieve your high-level goals.

robertgshaw2-neuralmagic commented:
@HandH1998 - where are these kernels from? Is there an open source library that we are tracking for these?

I know these are adapted from Marlin so I think we could ramp up on them quickly, but I wanted to understand if this is something that we [Neural Magic + rest of the vLLM team] need to maintain or if we are just simply tracking a remote repo

zhyncs commented:
It's adapted from Marlin and modified as necessary.
https://github.com/HandH1998/vllm/blob/b4b677138732ac952be8dbc87e44a5a58dffeb73/csrc/quantization/qqq/qqq_gemm_kernel.cu#L2-L7

> need to maintain

yep.

> simply tracking a remote repo

There seems to be no need at the moment.

robertgshaw2-neuralmagic commented:
> It's adapted from Marlin and modified as necessary. https://github.com/HandH1998/vllm/blob/b4b677138732ac952be8dbc87e44a5a58dffeb73/csrc/quantization/qqq/qqq_gemm_kernel.cu#L2-L7
> 
> > need to maintain
> 
> yep.
> 
> > simply tracking a remote repo
> 
> There seems to be no need at the moment.

Im just trying to understand who is responsible for maintaining these kernels moving forward?

We currently have two models
- With Punica / Flashattention / AWQ , we track from a different repo maintained by a 3rd party (+ we pull them into vllm as needed)
- With Marlin / cutlass / triton / paged_attention, we maintain them ourselves

So I am trying to understand which case this is

zhyncs commented:
> which case this is

I think it should be the latter.

On another note, speaking off the topic, [the performance of AWQ in vLLM is almost the worst](https://www.bentoml.com/blog/benchmarking-llm-inference-backends) among several mainstream frameworks. I also do not recommend continuing to use 3rd party, but rather maintaining our own set.

alexm-neuralmagic commented on csrc/quantization/qqq/qqq_gemm_kernel.cu:
```diff
@@ -0,0 +1,1318 @@
+      "l"(glob_ptr), "n"(BYTES));
+}
+
+// NOTE(HandH1998): cp.async.cg only support BYTES = 16, however,
```
Thanks for documenting this! It is helpful

alexm-neuralmagic commented on csrc/quantization/qqq/qqq_gemm_kernel.cu:
```diff
@@ -0,0 +1,1318 @@
+  // thread 0, 1, 2, 3. For more details, refer to mma operand A layout as s1's
+  // size is not fixed, we can not shuffle before inference we shuffle it when
+  // fetching s1 from global memory to shared memory, that's why s1_sh_wr is
+  // like this
```
Is "s1" statically defined with the quantization of the model or it is computed dynamically during inference?

HandH1998 replied:
The latter. Actually It is the dynamic per-token quantization scale, needing to be computed online.

alexm-neuralmagic commented on csrc/quantization/qqq/qqq_gemm_kernel.cu:
```diff
@@ -0,0 +1,1318 @@
+  FragC frag_c[thread_m_blocks][4][2];
+  FragS_GROUP frag_s3[2][4];
+  FragS_CHANNEL frag_s1[thread_m_blocks];
+  FragS_CHANNEL frag_s2[2][4];
```
During execution, the code uses only FragS_GROUP o FragS_CHANNEL. You can use if (constexpr) to avoid always defining both, or you are relying on the compiler to eliminate them anyway?

HandH1998 replied:
Actually, it only uses FragS_CHANNEL for per-channel dequantization, while it needs both FragS_GROUP and FragS_CHANNEL for per-group weight quantization. I will consider your suggestion.

HandH1998 replied:
Hi, I am trying to define  the register array `frag_s3` in `if constexpr`, but it is invisible to external variables. Do you know a method to solve it?

alexm-neuralmagic commented on csrc/quantization/qqq/qqq_gemm_kernel.cu:
```diff
@@ -0,0 +1,1318 @@
+        frag_b1 = dequant_per_group(b_quant_shift, frag_s3[k % 2][j], 1);
+      } else {
+        int b_quant_shift = b_quant << 4;
+        frag_b0 = dequant_per_channel(b_quant);
```
Why did you need to apply "dequant_per_channel" here and not on the final result (as it is done in the original Marlin, before write to C)?

alexm-neuralmagic replied:
I think I understand why you did it this way. You actually have a scale applied at the end on the output, and here you simply convert 4 bit to 8 bit by extracting bits. But maybe I'm missing something.

HandH1998 replied:
You are right. Here `dequant_per_channel` just converts INT4 to INT8 by left shifting 4 bits.

alexm-neuralmagic commented on csrc/quantization/qqq/qqq_gemm_kernel.cu:
```diff
@@ -0,0 +1,1318 @@
+  // finally have to globally reduce over the results. As the striped partioning
+  // minimizes the number of such reductions and our outputs are usually rather
+  // small, we perform this reduction serially in L2 cache.
+  auto global_reduce = [&](bool first = false, bool last = false) {
```
Would be good to document here that the global_reduce works on int32 elements (since the original marlin was half), and mention also that this is the reason that you needed to add the temporary buffer C (since the original Marlin was reducing directly on the output buffer)

HandH1998 replied:
I will fix it later.

alexm-neuralmagic commented on csrc/quantization/qqq/qqq_gemm_kernel.cu:
```diff
@@ -0,0 +1,1318 @@
+const int USER_THREADS =
+    256;               // Note: This is only used with user-provided thread_k/n
+const int STAGES = 4;  // 4 pipeline stages fit into shared memory
+const int SHARED_MEM =
```
You can change this code to dynamically detect the L1 cache size (like we did in gptq marlin and marlin24). This will provide you a better estimate of L1 cache, since sometimes you will need more L1 due to scales or increased batching.

HandH1998 replied:
Good suggestion! I will modify it.

robertgshaw2-neuralmagic commented:
> > which case this is
> 
> I think it should be the latter.
> 
> On another note, speaking off the topic, [the performance of AWQ in vLLM is almost the worst](https://www.bentoml.com/blog/benchmarking-llm-inference-backends) among several mainstream frameworks. I also do not recommend continuing to use 3rd party, but rather maintaining our own set.

sg - we are reviewing

re: AWQ. I think we can run AWQ models with Marlin if we support zero points in Marlin. We would welcome a contribution

alexm-neuralmagic commented on csrc/quantization/qqq/qqq_gemm_kernel.cu:
```diff
@@ -0,0 +1,1318 @@
+  __CALL_IF(4, N_BLOCKS, K_BLOCKS, -1, NUM_THREADS) \
+  __CALL_IF(4, N_BLOCKS, K_BLOCKS, 8, NUM_THREADS)
+
+void qqq_cuda(const void* A, const void* B, void* C, void* D, void* s1,
```
could you rename it to "marlin_qqq_cuda", I think it is better describing what is happening there.

HandH1998 replied:
OK.

alexm-neuralmagic commented on csrc/quantization/qqq/qqq_gemm_kernel.cu:
```diff
@@ -0,0 +1,1318 @@
+}  // namespace qqq
+}  // namespace vllm
+
+torch::Tensor qqq_gemm(torch::Tensor& a, torch::Tensor& b_q_weight,
```
same here, "marlin_qqq_gemm" is a better name in my opinion

HandH1998 replied:
OK.

alexm-redhat reviewed (commented):
Reviewed the marlin modified kernel, great work guys! Left some comments.

robertgshaw2-neuralmagic commented:
Another thing - what is the model serialization format?

zhyncs commented:
> Another thing - what is the model serialization format? And how do we make models in this format?

https://github.com/HandH1998/QQQ/blob/4ab12906bb144ca8977a077f7f191e218fc2e038/examples/quant_model.py#L76-L83

https://github.com/handh1998/qqq?tab=readme-ov-file#quantize-model

Due to [the changes](https://github.com/vllm-project/vllm/pull/5218/commits/b4b677138732ac952be8dbc87e44a5a58dffeb73) from the original pull request, the performance will have a slight difference compared to the data in the image.

alexm-neuralmagic commented on csrc/quantization/qqq/qqq_gemm_kernel.cu:
```diff
@@ -0,0 +1,1318 @@
+// Efficiently dequantize an int32 value into a full B-fragment of 4 int8 values
+// for weight per channel dequant.
+__device__ inline FragB dequant_per_channel(int q) {
+  static constexpr int MASK = 0xf0f0f0f0;
```
Isn't the mask supposed to be 0x0f0f0f0f? why do you extract the higher 4 bits?

HandH1998 replied:
For per-channel weight dequantization, I position the INT4 weight into the upper 4 bits of an INT8 by multiplying by 16, essentially performing a left shift by 4 bits. I use the mentioned mask and  line 624 to achieve it. Note that    the 8 INT4 weights  of one INT32 are shuffled offline to ensure fetching the correct weights for every thread of a wrap.

zhyncs commented:
Hi @alexm-neuralmagic These are the illustrations in the paper on QQQ drawn by @HandH1998 , hoping to help you review the code. The paper will be published on arxiv soon, please stay tuned.

<img src="https://github.com/vllm-project/vllm/assets/46627482/a0a50aaa-0347-47e7-a49b-53987c12471f" width="400" height="400">

<img src="https://github.com/vllm-project/vllm/assets/46627482/dabac1fd-653b-4eff-a190-2599d0924aff" width="400" height="650">

<img src="https://github.com/vllm-project/vllm/assets/46627482/dca596ff-ee02-402d-9a23-b0ae53593021" width="400" height="500">

<img src="https://github.com/vllm-project/vllm/assets/46627482/6cdb5e2a-70ba-4c38-829d-37b7db39002c" width="400" height="500">

alexm-neuralmagic commented:
@zhyncs thanks for the figure, this is really helpful

HandH1998 commented on csrc/quantization/qqq/qqq_gemm_kernel.cu:
```diff
@@ -0,0 +1,1318 @@
+// may be evicted immediately; used for quantized weights B, which are only
+// accessed precisely once and should thus not pollute the L2 cache which we
+// need for inputs A and outputs C.
+__device__ inline void cp_async4_stream(void* smem_ptr, const void* glob_ptr) {
```
I will try to modify functions `cp_async4_stream` and `cp_async1_stream` as you suggest next week. I think it should be easy to do it.

HandH1998 commented:
> So I am trying to understand which case this is

You guys can maintain qqq_gemm like you handle marlin. And remember to keep the copyright information at [https://github.com/HandH1998/vllm/blob/b4b677138732ac952be8dbc87e44a5a58dffeb73/csrc/quantization/qqq/qqq_gemm_kernel.cu#L2-L7](url). If I have other optimization ideas, I will let you know.

HandH1998 commented:
@robertgshaw2-neuralmagic @alexm-neuralmagic We have fixed the most issues you mentioned. For more details, please refer to our [paper](https://arxiv.org/pdf/2406.09904).

HandH1998 commented:
@robertgshaw2-neuralmagic @alexm-neuralmagic We have rebased our code on vllm 0.5.0 and hope you guys can review it.

tlrmchlsmth commented on csrc/quantization/qqq/qqq_gemm_kernel.cu:
```diff
@@ -0,0 +1,1317 @@
+          transform_a(a_sh_rd_delta_o * i + a_sh_rd_delta_i * j + a_sh_rd);
+  }
+
+  // Since B-accesses have non-constant stride they have to be computed at
+  // runtime; we break dependicies between subsequent accesses with a tile by
+  // maintining multiple pointers (we have enough registers), a tiny
+  // optimization.
```
nit: dependicies -> dependencies and maintining -> maintaining

I think codespell should catch this, so make sure you can run `format.sh` on this to get the build green :)

tlrmchlsmth reviewed (commented):
Could you add some unit tests for this kernel?

HandH1998 commented:
@tlrmchlsmth We are adding the GEMM unit test and fixing the issues. We will release the code next Monday.

zhyncs commented:
Hi @robertgshaw2-neuralmagic @alexm-neuralmagic @tlrmchlsmth May you help review this latest code? Thanks. If everything goes smoothly, after this PR is merged, out team @HandH1998 plan to continue working on INT4-FP8 QQQ (W4A8) based on this.

HandH1998 commented:
@tlrmchlsmth We have added w4a8 GEMM unit test and fixed review issues.

tlrmchlsmth commented on vllm/model_executor/layers/quantization/qqq.py:
```diff
@@ -0,0 +1,279 @@
+
+MARLIN_QQQ_SUPPORTED_NUM_BITS = [4]
+MARLIN_QQQ_SUPPORTED_GROUP_SIZES = [-1, 128]
+MARLIN_SUPPORTED_SYM = [True]
```
This isn't used anywhere. Should it be removed?

zhyncs replied:
ok

HandH1998 replied:
@tlrmchlsmth They are used in [vllm/model_executor/layers/quantization/qqq.py](https://github.com/vllm-project/vllm/pull/5218/commits/f8fba5ea74bfda8c99d904192c3ede3543377025#diff-bb4b60166935dd924ca19794d6af2c83cc115a5d0bbc0bb4b53ad9fdf5b54150R38-R49) and [tests/kernels/test_marlin_gemm.py](https://github.com/vllm-project/vllm/pull/5218/commits/f8fba5ea74bfda8c99d904192c3ede3543377025#diff-7cc4a2d906e22a9eace8c90f99dec52769bfc5a884ef22558f5ace310ed929a9R229-R230). `vllm/model_executor/layers/quantization/gptq_marlin.py` did the similar thing. I think it should be retained.

tlrmchlsmth commented on csrc/quantization/qqq/qqq_gemm_kernel.cu:
```diff
@@ -0,0 +1,1239 @@
+      *reinterpret_cast<half2*>(&t1), *reinterpret_cast<const half2*>(&MUL),
+      *reinterpret_cast<const half2*>(&ADD));
+
+  uint16_t s = reinterpret_cast<uint16_t*>(&frag_s)[i];
+  uint32_t double_s;
+  // pack 2xfp16 to half2
+  asm volatile("mov.b32 %0, {%1, %2};\n" : "=r"(double_s) : "h"(s), "h"(s));
+  // dequant and convert 4 half to 4 uint8 (be placed at the low 8 bits of 4
+  // half, respectively)
+  static constexpr uint32_t MAGIC_NUM = 0x64806480;
+  *reinterpret_cast<half2*>(&t0) = __hfma2(
+      *reinterpret_cast<half2*>(&t0), *reinterpret_cast<half2*>(&double_s),
+      *reinterpret_cast<const half2*>(&MAGIC_NUM));
+  *reinterpret_cast<half2*>(&t1) = __hfma2(
+      *reinterpret_cast<half2*>(&t1), *reinterpret_cast<half2*>(&double_s),
+      *reinterpret_cast<const half2*>(&MAGIC_NUM));
+  // take out the 4 uint8 from 4 half, then convert them to 4 int8 and pack 4
+  // int8 into 1 uint32
+  FragB frag_b;
+  uint32_t uint8s;
+  static constexpr uint32_t MASK_0246 = 0x6420;
+  static constexpr uint32_t UINT8s_TO_INT8s_MASK = 0x80808080;
+  asm volatile("prmt.b32 %0,%1,%2,%3;\n"
+               : "=r"(uint8s)
+               : "r"(t0), "r"(t1), "n"(MASK_0246));
+  frag_b[0] = (uint8s ^ UINT8s_TO_INT8s_MASK);
+  return frag_b;
```
I'm a little confused about something. In the paper it says:

> Initially, INT4 weights are converted to FP16 using the FastINT4toFP16 (Kim et al., 2022) method. Subsequently, each group's weight is dequantized by multiplying it by its respective scale. To re-quantize into INT8, the result is divided by the per-channel quantization scale. These two steps are elegantly merged into a single operation, FusedDequantQuant, by dividing the per-group scale by the per-channel scale, as illustrated in Figure 4. The final stage of the process involves converting the FP16 results of FusedDequantQuant back into INT8 format to proceed with the INT8 GEMM. Similar to the per-channel method, the INT32 output from the INT8 GEMM undergoes dequantization to FP16.

Where does dividing by the per-channel quantization scale happen?

zhyncs replied:
https://github.com/vllm-project/vllm/pull/5218/files#diff-f444886cd1e6af281a9a0a224813fb9c74424a2b889ecc050b4372c119f669f1R160-R172

tlrmchlsmth replied:
I see a multiplication with a value in `frag_s` there, which I thought contained the group's scale. Does the per-group scale get divided by the per-channel scale before this function gets called? What am I missing? Could you please explain?

HandH1998 replied:
The per-group scale is divided by the per-channel scale when exporting the w4a8 model offline in our repo [https://github.com/HandH1998/QQQ/blob/49f06e0b47c606ca2c5558ade0805b0609d57a8f/QQQ/gptq/qlinear/qlinear_marlin.py#L193-L195](url), as explained in the paper `These two steps are elegantly merged into a single operation, FusedDequantQuant, by dividing the per-group scale by the per-channel scale, as illustrated in Figure 4.`.

tlrmchlsmth replied:
I see, thanks for explaining.

tlrmchlsmth reviewed (commented):
I have a few questions/comments but the PR looks pretty good!

zhyncs commented on csrc/quantization/marlin/dense/marlin_cuda_kernel.cu:
```diff
@@ -25,30 +25,26 @@
 
 #include <iostream>
 
+#include "common/base.h"
+
+#if defined(__CUDA_ARCH__) && __CUDA_ARCH__ < 800
+
+#else
+
+  #include "common/mem.h"
+
+#endif
```
ok

zhyncs commented on csrc/quantization/qqq/qqq_gemm_kernel.cu:
```diff
@@ -0,0 +1,1239 @@
+  return std::to_string(x);
+}
+
+namespace qqq {
```
make sense

zhyncs commented on csrc/quantization/qqq/qqq_gemm_kernel.cu:
```diff
@@ -0,0 +1,1239 @@
+  static constexpr uint32_t HI = 0x00f000f0;
+  static constexpr uint32_t EX = 0x64006400;
+  // Guarantee that the `(a & b) | c` operations are LOP3s.
+  uint32_t t0 = lop3<(0xf0 & 0xcc) | 0xaa>(q, LO, EX);
+  uint32_t t1 = lop3<(0xf0 & 0xcc) | 0xaa>(q, HI, EX);
```
Do you mean `lop3<0xea>`?

tlrmchlsmth replied:
Yes you're right I made a mistake in my bitwise arithmetic -- but I still have the same general question.

zhyncs replied:
https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#logic-and-shift-instructions-lop3

tlrmchlsmth replied:
ok, pretty cool actually. And I now see that this is standard usage among the Marlin kernels :+1:

zhyncs commented on csrc/quantization/marlin/dense/common/base.h:
```diff
@@ -0,0 +1,31 @@
+/*
+ * Modified by HandH1998
```
ok

zhyncs commented on vllm/model_executor/layers/quantization/utils/marlin_utils.py:
```diff
@@ -57,6 +59,30 @@ def marlin_weights(q_w, size_k, size_n, num_bits, perm):
+    pack_factor = get_pack_factor(num_bits)
+    orig_device = q_w.device
+
+    q_w = q_w.cpu().numpy().astype(numpy.uint32)
+
+    q_packed = numpy.zeros((q_w.shape[0], q_w.shape[1] // pack_factor),
+                           dtype=numpy.uint32)
+    if group_size == size_k:
+        for i in range(pack_factor):
+            q_packed |= (q_w[:, i::pack_factor] & 0xF) << num_bits * i
+    else:
+        for i in range(pack_factor):
+            q_packed |= q_w[:, i::pack_factor] << num_bits * i
+
+    q_packed = torch.from_numpy(q_packed.astype(numpy.int32)).to(orig_device)
```
Not used online and not a bottleneck, optimization may not significant.

zhyncs commented:
This week, @HandH1998 is focusing on implementing QQQ in LMDeploy. We expect to address these valuable review feedbacks you mentioned and resolve conflicts this week. Please stay tuned. @tlrmchlsmth @mgoin

HandH1998 commented on vllm/model_executor/layers/quantization/qqq.py:
```diff
@@ -12,46 +12,64 @@
-                "Currently, only group size 128 and -1 (channelwise) "
-                "is supported for QQQ, but got group_size of "
-                f"{self.group_size}")
+                f"QQQ does not support group_size = {self.group_size}. "
+                f"Only group_sizes = {MARLIN_QQQ_SUPPORTED_GROUP_SIZES} "
+                "are supported.")
```
ok

haichuan1221 commented:
excellent work! I checked QQQ-g128 and QoQ-g128, it seems that QoQ is much faster on A100/A6000 devices, is this true?

HandH1998 commented:
> excellent work! I checked QQQ-g128 and QoQ-g128, it seems that QoQ is much faster on A100/A6000 devices, is this true?

QoQ is an excellent work which employs many techniques to improve accuracy. 
However, QoQ works in qserver, while QQQ is integrated into vllm. Comparing inference performance across different inference systems may not be entirely fair.
Additionally, it appears that qserver consistently enables kv4/8, except for w4a8, and kv cache quantization can also impact performance. This factor should be taken into account as well.
Since QoQ hasn't released their w4a8 GEMM kernel code, I am unable to determine which implementation is more efficient.
Hope this is helpful to you.

haichuan1221 commented:
> > excellent work! I checked QQQ-g128 and QoQ-g128, it seems that QoQ is much faster on A100/A6000 devices, is this true?
> 
> QoQ is an excellent work which employs many techniques to improve accuracy. However, QoQ works in qserver, while QQQ is integrated into vllm. Comparing inference performance across different inference systems may not be entirely fair. Additionally, it appears that qserver consistently enables kv4/8, except for w4a8, and kv cache quantization can also impact performance. This factor should be taken into account as well. Since QoQ hasn't released their w4a8 GEMM kernel code, I am unable to determine which implementation is more efficient. Hope this is helpful to you.

Indeed, QoQ use 4 bit kv cache, do you have any plan to support 4 bit kv cache quantization in the future?

zhyncs commented:
> have any plan to support 4 bit kv cache

There are no plans for this in the near future. We are currently working on integrating QQQ with LMDeploy. The follow-up plan is to continue working on INT4-FP8 QQQ (W4A8) based on this.

HandH1998 commented:
> > > excellent work! I checked QQQ-g128 and QoQ-g128, it seems that QoQ is much faster on A100/A6000 devices, is this true?
> > 
> > 
> > QoQ is an excellent work which employs many techniques to improve accuracy. However, QoQ works in qserver, while QQQ is integrated into vllm. Comparing inference performance across different inference systems may not be entirely fair. Additionally, it appears that qserver consistently enables kv4/8, except for w4a8, and kv cache quantization can also impact performance. This factor should be taken into account as well. Since QoQ hasn't released their w4a8 GEMM kernel code, I am unable to determine which implementation is more efficient. Hope this is helpful to you.
> 
> Indeed, QoQ use 4 bit kv cache, do you have any plan to support 4 bit kv cache quantization in the future?

I do not plan to support kv4 in QQQ, as QQQ primarily focuses on weight quantization. However, I think that vllm will consider supporting kv4 in the future as more advanced kv4 methods become available.

haichuan1221 commented:
> > have any plan to support 4 bit kv cache
> 
> There are no plans for this in the near future. We are currently working on integrating QQQ with LMDeploy. The follow-up plan is to continue working on INT4-FP8 QQQ (W4A8) based on this.

Thank you for your reply and hope to see int4-fp8 performace in near future

zhyncs commented:
@haichuan1221 And by the way, according to [our previous experience](https://github.com/InternLM/lmdeploy/releases/tag/v0.4.0), KV Cache Int4 has a greater impact on accuracy compared to KV Cache Int8 and fp16. We basically do not use KV Cache Int4 online.

brisker commented:
@zhyncs 
@HandH1998 
it seems that w4a8(no group) is much faster than w4a8(with group), but worse accuracy. Any plan on combining w4a8 and w4a8gs128 together, to achieve accuracy-efficiency trade-off?

Hongbosherlock commented:
> @haichuan1221 And by the way, according to [our previous experience](https://github.com/InternLM/lmdeploy/releases/tag/v0.4.0), KV Cache Int4 has a greater impact on accuracy compared to KV Cache Int8 and fp16. We basically do not use KV Cache Int4 online.

![image](https://github.com/vllm-project/vllm/assets/27813268/dfc9acec-dc60-4e8f-b218-1daed0a4c060)

Hi @zhyncs When the `number of input tokens` is 1, the speedup of W4A8 is `3.5`, and the speedup of W8A8 is `1.5`. The former is more than **\**twice\**** the performance of the latter. 
Can you explain why? Thanks!

zhyncs commented:
Hi @robertgshaw2-neuralmagic @alexm-neuralmagic @tlrmchlsmth @mgoin 
@HandH1998 has resolved the review comments, may you help review the latest code? Thanks.

brisker commented:
@HandH1998 
Regarding the difference between w4a8 and w4a8-fusion, why the w4a8-fusion version can not support all models?(since the latter on is faster, why not use the latter one? I think dynamic act quantization can not be a big problem.)
![image](https://github.com/vllm-project/vllm/assets/13804492/90328ba6-ffd6-4180-9f5a-d9b0638de36b)

zhyncs commented:
> Regarding the difference between w4a8 and w4a8-fusion, why the w4a8-fusion version can not support all models?(since the latter one is faster, why not use the latter one? I think dynamic act quantization can not be a big problem.)

ref https://github.com/vllm-project/vllm/pull/5218#issuecomment-2145446331

haichuan1221 commented:
I run qqq with vllm and met the error: 
`Exception` raised from c10_cuda_check_implementation at ../c10/cuda/CUDAException.cpp:43 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::string) + 0x57 (0x7f6d10e48897 in /mnt/harddisk/miniconda3/envs/vllm_w4a8/lib/python3.9/site-packages/torch/lib/libc10.so)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::string const&) + 0x64 (0x7f6d10df8b25 in /mnt/harddisk/miniconda3/envs/vllm_w4a8/lib/python3.9/site-packages/torch/lib/libc10.so)
frame #2: c10::cuda::c10_cuda_check_implementation(int, char const*, char const*, int, bool) + 0x118 (0x7f6d10f20718 in /mnt/harddisk/miniconda3/envs/vllm_w4a8/lib/python3.9/site-packages/torch/lib/libc10_cuda.so)
frame #3: c10d::ProcessGroupNCCL::WorkNCCL::finishedGPUExecutionInternal() const + 0x56 (0x7f6d1211ee36 in /mnt/harddisk/miniconda3/envs/vllm_w4a8/lib/python3.9/site-packages/torch/lib/libtorch_cuda.so)
frame #4: c10d::ProcessGroupNCCL::WorkNCCL::isCompleted() + 0x58 (0x7f6d12122f38 in /mnt/harddisk/miniconda3/envs/vllm_w4a8/lib/python3.9/site-packages/torch/lib/libtorch_cuda.so)
frame #5: c10d::ProcessGroupNCCL::watchdogHandler() + 0x77c (0x7f6d121285ac in /mnt/harddisk/miniconda3/envs/vllm_w4a8/lib/python3.9/site-packages/torch/lib/libtorch_cuda.so)
frame #6: c10d::ProcessGroupNCCL::ncclCommWatchdog() + 0x10c (0x7f6d1212931c in /mnt/harddisk/miniconda3/envs/vllm_w4a8/lib/python3.9/site-packages/torch/lib/libtorch_cuda.so)
frame #7: <unknown function> + 0xdbbf4 (0x7f6d5dbdbbf4 in /mnt/harddisk/miniconda3/envs/vllm_w4a8/bin/../lib/libstdc++.so.6)
frame #8: <unknown function> + 0x8609 (0x7f6d5ee9e609 in /lib/x86_64-linux-gnu/libpthread.so.0)
frame #9: clone + 0x43 (0x7f6d5ec69353 in /lib/x86_64-linux-gnu/libc.so.6)

[rank0]:[E ProcessGroupNCCL.cpp:1414] [PG 2 Rank 0] Process group watchdog thread terminated with exception: CUDA error: an illegal memory access was encountered
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Compile with `TORCH_USE_CUDA_DSA` to enable device-side ` 

It seems there might be some issue related to memory access

Hongbosherlock commented:
> ![image](https://private-user-images.githubusercontent.com/27813268/345405487-dfc9acec-dc60-4e8f-b218-1daed0a4c060.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAxNjg5OTYsIm5iZiI6MTcyMDE2ODY5NiwicGF0aCI6Ii8yNzgxMzI2OC8zNDU0MDU0ODctZGZjOWFjZWMtZGM2MC00ZThmLWIyMTgtMWRhZWQwYTRjMDYwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA1VDA4MzgxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRjNzc4ZmIyMTI0ZmJlODQ2ZTVjMThhMTYxYWRjODE0ZmU1ZmI4NTFiMDBmMDg4NmUzNGE4ZmUxMjFmODk3NWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Jk3nM1F_4fnNcFmGbuAOghdxIooR4NCjjCsJs84KiFA)
> 
> Hi @zhyncs When the `number of input tokens` is 1, the speedup of W4A8 is `3.5`, and the speedup of W8A8 is `1.5`. The former is more than ***_twice*_** the performance of the latter. Can you explain why? Thanks!

Hi @tlrmchlsmth , when profiling a single GEMM, like the picture above shows. Do you think the performance of `W4W8` would be **2x** that of [W8A8](https://github.com/vllm-project/vllm/pull/4749) here?

tlrmchlsmth commented:
I would also like to try this PR out before accepting -- Is there a QQQ model on huggingface? If not could you please upload one (ideally a llama3-8b)? 

I did try to create my own QQQ model but ran into an issue with a hard coded file path, and ideally users should have an easy path to use this kernel before landing.

zhyncs commented:
Hi @tlrmchlsmth We went on a team building activity today. We plan to upload tomorrow, so please stay tuned. cc @HandH1998

HandH1998 commented:
@tlrmchlsmth I have uploaded quantized Llama-3-8b models to hf repo [QQQ-Llama-3-8b](https://huggingface.co/HandH1998/QQQ-Llama-3-8b) and [QQQ-Llama-3-8b-g128](https://huggingface.co/HandH1998/QQQ-Llama-3-8b-g128). And I also released w4a8 kernel [repo](https://github.com/HandH1998/marlin/tree/w4a8). You can use [bench_w4a8.py](https://github.com/HandH1998/marlin/blob/w4a8/bench_w4a8.py) to benchmark the w4a8 performance and use [test_w4a8.py](https://github.com/HandH1998/marlin/blob/w4a8/test_w4a8.py) to verify the correctness.

tlrmchlsmth commented:
Hey @HandH1998, I am running into an issue with both QQQ Llama3 models. As a simple sanity check, I've modified `examples/offline_inference` as follows:
```
diff --git a/examples/offline_inference.py b/examples/offline_inference.py
index 9b758fa2..f5c14f40 100644
--- a/examples/offline_inference.py
+++ b/examples/offline_inference.py
@@ -11,7 +11,7 @@ prompts = [
 sampling_params = SamplingParams(temperature=0.8, top_p=0.95)

 # Create an LLM.
-llm = LLM(model="facebook/opt-125m")
+llm = LLM(model="HandH1998/QQQ-Llama-3-8b-g128")
 # Generate texts from the prompts. The output is a list of RequestOutput objects
 # that contain the prompt, generated text, and other information.
 outputs = llm.generate(prompts, sampling_params)
```

and the generated text is garbage:
```
Prompt: 'Hello, my name is', Generated text: ' Bast bathroomubitUsageId(rowsrupズoticotel Hut_loginQAviouslycht.Collections contemplate'
Prompt: 'The president of the United States is', Generated text: ' ldc Lottoanáagation succession �atile Gouldonus LastorealachtenUtcBlank_mapsروج'
Prompt: 'The capital of France is', Generated text: 'ину;amp Educational NotreикуχριamicweeneyFore cr Lorenzoesianamb Roechuremenin'
Prompt: 'The future of AI is', Generated text: 'jav.quest.googleapis smemd[Any sco VegannostICLE पह empirečnéOBJ puls mov'
```

Any idea what might be going wrong? I'm on an A100

HandH1998 commented:
@tlrmchlsmth It should be modified to `llm = LLM(model="HandH1998/QQQ-Llama-3-8b-g128", quantization="qqq')`. You should specify the quantization method.

tlrmchlsmth commented:
@HandH1998 that works, thanks. I think it'd be nice to get that autodetected based on the model's `quant_config.json`. I'm not familiar with that part of the model loading but maybe @mgoin or @dsikka can help there

tlrmchlsmth commented:
@HandH1998 Update: I see you have a quant_config.json: https://huggingface.co/HandH1998/QQQ-Llama-3-8b-g128/blob/main/quant_config.json. This was used at some point for AutoGPTQ/AWQ, but we have since moved these to be within config.json.

Then for this model to work out of the box without users needing to specify `quantization="qqq"`, all you need to do is add an entry to your config.json like:
```
  "quantization_config": {
    "group_size": 128,
    "quant_method": "qqq"
    "wbits": 4,
  },
```

HandH1998 commented:
@tlrmchlsmth Got it. I will fix it later.

HandH1998 commented:
> all you need to do is add an entry to your config.json like:
> 
> ```
>   "quantization_config": {
>     "group_size": 128,
>     "quant_method": "qqq",
>     "wbits": 4
>   },
> ```

I have fixed it. Now the quantized model exported by QQQ repo will put quant_config into the entry `quantization_config` of model's `config.json`. And I also updated hf repo [QQQ-Llama-3-8b](https://huggingface.co/HandH1998/QQQ-Llama-3-8b) and [QQQ-Llama-3-8b-g128](https://huggingface.co/HandH1998/QQQ-Llama-3-8b-g128).

tlrmchlsmth commented:
I ran some evals on my end with the following command. Do these look reasonable to you all?

```
lm_eval --model vllm \
    --model_args pretrained=$model,quantization=qqq,dtype=auto,gpu_memory_utilization=0.6 \
    --tasks hellaswag,lambada_openai,arc_easy \
    --batch_size auto \
    --limit 1000
```
#### HandH1998/QQQ-Llama-3-8b
|    Tasks     |Version|Filter|n-shot|  Metric  |   |Value |   |Stderr|
|--------------|------:|------|-----:|----------|---|-----:|---|-----:|
|arc_easy      |      1|none  |     0|acc       |↑  |0.7480|±  |0.0137|
|              |       |none  |     0|acc_norm  |↑  |0.7100|±  |0.0144|
|hellaswag     |      1|none  |     0|acc       |↑  |0.4980|±  |0.0158|
|              |       |none  |     0|acc_norm  |↑  |0.6520|±  |0.0151|
|lambada_openai|      1|none  |     0|acc       |↑  |0.6830|±  |0.0147|
|              |       |none  |     0|perplexity|↓  |4.6465|±  |0.2193|

#### HandH1998/QQQ-Llama-3-8b-g128
|    Tasks     |Version|Filter|n-shot|  Metric  |   |Value|   |Stderr|
|--------------|------:|------|-----:|----------|---|----:|---|-----:|
|arc_easy      |      1|none  |     0|acc       |↑  |0.767|±  |0.0134|
|              |       |none  |     0|acc_norm  |↑  |0.744|±  |0.0138|
|hellaswag     |      1|none  |     0|acc       |↑  |0.514|±  |0.0158|
|              |       |none  |     0|acc_norm  |↑  |0.679|±  |0.0148|
|lambada_openai|      1|none  |     0|acc       |↑  |0.745|±  |0.0138|
|              |       |none  |     0|perplexity|↓  |3.341|±  |0.1446|

#### meta-llama/Meta-Llama-3-8B
|    Tasks     |Version|Filter|n-shot|  Metric  |   |Value |   |Stderr|
|--------------|------:|------|-----:|----------|---|-----:|---|-----:|
|arc_easy      |      1|none  |     0|acc       |↑  |0.8050|±  |0.0125|
|              |       |none  |     0|acc_norm  |↑  |0.7660|±  |0.0134|
|hellaswag     |      1|none  |     0|acc       |↑  |0.5250|±  |0.0158|
|              |       |none  |     0|acc_norm  |↑  |0.6870|±  |0.0147|
|lambada_openai|      1|none  |     0|acc       |↑  |0.7740|±  |0.0132|
|              |       |none  |     0|perplexity|↓  |3.0714|±  |0.1249|`

tlrmchlsmth reviewed (commented):
Do you have any small QQQ w4a8 models we could use for integration testing? Something like a QQQ OPT-125m would be ideal here. Without an integration test, something will eventually break

tlrmchlsmth commented on csrc/quantization/qqq/qqq_gemm_kernel.cu:
Do you think we should rename qqq_gemm_kernel.cu to marlin_qqq_gemm_kernel.cu and put it under quantization/marlin/qqq directory? Then it will be clear that it is part of the family of marlin extensions

HandH1998 replied:
ok

HandH1998 commented on csrc/quantization/qqq/qqq_gemm_kernel.cu:
```diff
@@ -0,0 +1,1233 @@
+                           // (max_par*16*4)xn, as int8 tensor core's output is
+                           // int32 dtype
+    int4* __restrict__ D,          // fp16 output buffer of shape mxn
+    const float* __restrict__ s1,  // fp32 activation per-token quantization
+                                   // scales of shape mx1
+    const int4* __restrict__ s2,  // fp32 weight per-channel quantization scales
+                                  // of shape 1xn
+    const int4* __restrict__ s3,  // fp16 weight per-group quantization scales
```
ok

HandH1998 commented:
> Do you have any small QQQ w4a8 models we could use for integration testing? Something like a QQQ OPT-125m would be ideal here. Without an integration test, something will eventually break

HandH1998 commented:
> I ran some evals on my end with the following command. Do these look reasonable to you all?
> 
> ```
> lm_eval --model vllm \
>     --model_args pretrained=$model,quantization=qqq,dtype=auto,gpu_memory_utilization=0.6 \
>     --tasks hellaswag,lambada_openai,arc_easy \
>     --batch_size auto \
>     --limit 1000
> ```

The results look reasonable to me. We plan to employ rotation method like `Quarot` to improve quantization performance, especially for per-channel. This will have no affect on inference speed and the current PR code, and I will remind you guys once the work is done.

HandH1998 commented:
> Do you have any small QQQ w4a8 models we could use for integration testing? Something like a QQQ OPT-125m would be ideal here. Without an integration test, something will eventually break

Sorry, QQQ now only supports Llama series, and the smallest model is QQQ-Llama-2-7b. We now still focus on the improving model performance. Support for more models will be put off to a later time. Can the QQQ-LLama-2-7b model be used for the integration testing? If you need, I will upload to the hf repo.

tlrmchlsmth commented:
>The results look reasonable to me. We plan to employ rotation method like Quarot to improve quantization performance, especially for per-channel. This will have no affect on inference speed and the current PR code, and I will remind you guys once the work is done.

Cool! Looking forward to it. 

For the integration test, yes let's get that QQQ-Llama-2-7b hooked up to the lm-evaluation-harness.

* First, run a baseline using this script:
https://github.com/vllm-project/vllm/blob/main/.buildkite/lm-eval-harness/run-lm-eval-gsm-vllm-baseline.sh
* Then add config scores like this:
https://github.com/vllm-project/vllm/blob/main/.buildkite/lm-eval-harness/configs/Meta-Llama-3-8B-Instruct.yaml
* Last add the name of the config to this file:
https://github.com/vllm-project/vllm/blob/main/.buildkite/lm-eval-harness/configs/models-small.txt

Then automation will pick it up. Then, once you address my last couple of minor review comments, it's good to land from my perspective. Thanks!

HandH1998 commented:
@tlrmchlsmth I have fixed your review comments and added Llama-3-8b to integration test  [https://github.com/vllm-project/vllm/pull/5218/commits/568f093f768b8865f7be03d36ff7c10215a27cf4#diff-a017cc0508457c9b1c9e57f62a5ce5ab72573b873e19b8c5d56e161b9844bc6cR1-R11](url).

brisker commented:
@HandH1998 
@zhyncs 
I got a w4a8 llama2-7b-model with **wiki-ppl=5.7** when tested using fake-quant, but when I convert the weight into qqq-w4a8 format and test it again using the cuda-kernel, **the wiki-ppl becomes 10000.** So I run the following code to test the consistency between fake-quant and qqq-w4a8-kernel, I find there seems to be some difference which **can not be neglected.**
The `QuantLinear `class is extracted from [here](https://github.com/HandH1998/QQQ/blob/main/QQQ/gptq/qlinear/qlinear_marlin.py#L46)

Any advice on this? I think this may be a potential bug.

```
from copy import deepcopy
import torch
import torch.nn as nn
import sys 
import QuantLinear

def dynamic_quant(x: torch.Tensor):
    quant_scale = x.abs().max(dim=-1, keepdim=True)[0].div(127.0).to(torch.float)
    quant_x = (x / quant_scale).round().clamp(-128, 127).to(torch.int8)
    dequant_x = quant_x*quant_scale
    return quant_x, quant_scale,dequant_x

def get_scale(x):
    q_max = (2 ** (4-1)) - 1
    x_absmax = x.abs().amax(dim=-1, keepdim=True)

    scale = x_absmax / q_max

    return scale,None


def w4_quant(x):
    q_max = (2 ** (4-1)) - 1
    x_absmax = x.abs().amax(dim=-1, keepdim=True)

    scale = x_absmax / q_max


    quant_value = torch.clamp(torch.round(x / scale), -q_max, q_max)

    x_dequant = quant_value * scale

    return x_dequant
ori_fc = nn.Linear(2048,4096,bias=False).cuda().half()

qqq_linear = QuantLinear(
                4,
                -1,
                2048,
                4096,
                False,
                weight_dtype=torch.float16
            )
qqq_linear.cuda()

input = torch.Tensor(10,2048).normal_().cuda().half()
quant_input, quant_input_scale,dequant_input = dynamic_quant(input)


scale,scale_extra = get_scale(ori_fc.weight.data)
qqq_linear.pack(ori_fc, scale, scale_extra)
ori_fc.weight.data = w4_quant(ori_fc.weight.data)

with torch.no_grad():
    out1 = ori_fc(dequant_input.half())

with torch.no_grad():
    out2 = qqq_linear.forward(input)

print((out1==out2).sum(),out1.numel())
```

HandH1998 commented:
> print((out1==out2).sum(),out1.numel())

Why you print this? I think `(out1 - out2).abs().sum()/out1.numel()` may be a good metric to evaluate the diff.
I ran your script, and the result seems normal.
![image](https://github.com/user-attachments/assets/00af3c01-50e6-4b99-a3b9-01655f95dd34)

brisker commented:
@HandH1998 

1.  why can not fake-quant be 100% the same?
2.  in your qqq-paper, are all the experiments and accuracy of w4a8 tested on cuda-kernel or fake-quant?

HandH1998 commented:
@brisker 
1. The order of underlying calculations will affect the results. Obviously, Obviously, these two have different calculation orders.
2. We used cuda-kernel `qqq_gemm.cu`.
3. Never found large diffs.

brisker commented:
@HandH1998 
very weird thing is that, if you modify
`x_absmax = x.abs().amax(dim=-1, keepdim=True)`
to
`x_absmax = x.abs().amax(dim=-1, keepdim=True) * 0.4999`
in the `get_scale` and `w4_quant` function in my script, **large diff occurs.**

I just tried it, and found this.


```
out1
tensor([[-0.0216,  0.5732,  0.4524,  ..., -0.4541,  0.2771, -0.4653],
        [-0.0375, -0.0845,  0.6333,  ...,  0.2949,  0.4790,  0.6089],
        [-0.1888, -0.1674, -0.4871,  ...,  0.2385,  0.7886,  0.2507],
        ...,
        [-0.0891,  0.2343,  0.1569,  ...,  0.6831, -0.1757, -0.5054],
        [-0.4187,  0.5977, -0.0225,  ...,  0.3330,  0.1484, -0.7993],
        [-0.4817, -0.1526, -0.1072,  ...,  0.4109, -0.0023, -0.5581]],
       device='cuda:0', dtype=torch.float16)
out2
tensor([[ 0.1066, -0.5508, -0.7207,  ...,  0.5381, -0.3894,  0.0118],
        [ 0.0127, -0.4534, -0.3826,  ..., -0.1334, -0.0036, -0.2688],
        [-0.1793, -0.0451,  0.4478,  ...,  0.4104,  0.6265,  0.5024],
        ...,
        [ 0.0429,  0.4233,  0.3572,  ..., -0.1161, -0.4062, -0.0836],
        [ 0.2568,  0.0115,  0.0422,  ...,  0.2424,  0.0394,  0.2054],
        [ 0.1392, -0.0248, -0.1234,  ..., -0.1874, -0.1741,  0.2067]],
       device='cuda:0', dtype=torch.float16)
```

tlrmchlsmth reviewed (approved):
@HandH1998 thanks for adding the integration test! PR looks good to me now.

I am going to go ahead and accept it, but before landing it, let's run down the issue that @brisker ran into, just to make sure nothing funny is going on. @brisker could you share the result of vllm's collect_env.py script? I'd like to make sure you haven't uncovered some edge cases we're missing. The output of that script looks good to me on an A100. 

Also @brisker could you also try re-running your test script with 
```assert(torch.allclose(out1, out2, atol=1e-3, rtol=1e-3))```
That's going to be a better way to check if the implementations match.

brisker commented:
@HandH1998 
@tlrmchlsmth 
I have figured out why weird thing happens:

It turns out that, when packing the model, if quanti scales is absmax, nothing weird happens.

But if quanti scales is not absmax, for the w4a8 kernel in this PR, the `B` (quantized w for the kernel) must be calculated by the fake-quant weight(quant-dequant weight), and scales must be calculated by original fp16 weight.

So for the weird results I mentioned above, after modifying
`x_absmax = x.abs().amax(dim=-1, keepdim=True)`
into
`x_absmax = x.abs().amax(dim=-1, keepdim=True)* 0.4999 `,   ( any (0,1) small numbers will all cause weird results, not only 0.4999)

if I modify further:

modify
```
scale,scale_extra = get_scale(ori_fc.weight.data)
qqq_linear.pack(ori_fc, scale, scale_extra)
ori_fc.weight.data = w4_quant(ori_fc.weight.data)
```
into
```
scale,scale_extra = get_scale(ori_fc.weight.data)
ori_fc.weight.data = w4_quant(ori_fc.weight.data)
qqq_linear.pack(ori_fc, scale, scale_extra)  
```

**then out1 and out2 becomes close again**


So maybe this is due to the special calculation pipeline for this w4a8 kernel.

But anyway, in my opinion, this is a little weird, because normally, we should  get quantized `B
`  and  `scales`  **all from the original fp16 weight.**

> @HandH1998 very weird thing is that, if you modify `x_absmax = x.abs().amax(dim=-1, keepdim=True)` to `x_absmax = x.abs().amax(dim=-1, keepdim=True) * 0.4999` in the `get_scale` and `w4_quant` function in my script, **large diff occurs.**
> 
> I just tried it, and found this.
> 
> ```
> out1
> tensor([[-0.0216,  0.5732,  0.4524,  ..., -0.4541,  0.2771, -0.4653],
>         [-0.0375, -0.0845,  0.6333,  ...,  0.2949,  0.4790,  0.6089],
>         [-0.1888, -0.1674, -0.4871,  ...,  0.2385,  0.7886,  0.2507],
>         ...,
>         [-0.0891,  0.2343,  0.1569,  ...,  0.6831, -0.1757, -0.5054],
>         [-0.4187,  0.5977, -0.0225,  ...,  0.3330,  0.1484, -0.7993],
>         [-0.4817, -0.1526, -0.1072,  ...,  0.4109, -0.0023, -0.5581]],
>        device='cuda:0', dtype=torch.float16)
> out2
> tensor([[ 0.1066, -0.5508, -0.7207,  ...,  0.5381, -0.3894,  0.0118],
>         [ 0.0127, -0.4534, -0.3826,  ..., -0.1334, -0.0036, -0.2688],
>         [-0.1793, -0.0451,  0.4478,  ...,  0.4104,  0.6265,  0.5024],
>         ...,
>         [ 0.0429,  0.4233,  0.3572,  ..., -0.1161, -0.4062, -0.0836],
>         [ 0.2568,  0.0115,  0.0422,  ...,  0.2424,  0.0394,  0.2054],
>         [ 0.1392, -0.0248, -0.1234,  ..., -0.1874, -0.1741,  0.2067]],
>        device='cuda:0', dtype=torch.float16)
> ```

brisker commented:
@HandH1998 
@tlrmchlsmth 
I have figured out why weird thing happens:

It turns out that, when packing the model, if quanti scales is absmax, nothing weird happens.

But if quanti scales is not absmax, for the w4a8 kernel in this PR, the `B` (quantized w for the kernel) must be calculated by the fake-quant weight(quant-dequant weight), and scales must be calculated by original fp16 weight.

So for the weird results I mentioned above, after modifying
`x_absmax = x.abs().amax(dim=-1, keepdim=True)`
into
`x_absmax = x.abs().amax(dim=-1, keepdim=True)* 0.4999 `,   ( any (0,1) small numbers will all cause weird results, not only 0.4999)

if I modify further:

modify
```
scale,scale_extra = get_scale(ori_fc.weight.data)
qqq_linear.pack(ori_fc, scale, scale_extra)
ori_fc.weight.data = w4_quant(ori_fc.weight.data)
```
into
```
scale,scale_extra = get_scale(ori_fc.weight.data)
ori_fc.weight.data = w4_quant(ori_fc.weight.data)
qqq_linear.pack(ori_fc, scale, scale_extra)  
```

**then out1 and out2 becomes close again**


So maybe this is due to the special calculation pipeline for this w4a8 kernel.

But anyway, in my opinion, this is a little weird, because normally, we should  get quantized `B
`  and  `scales`  **all from the original fp16 weight.**

> @HandH1998 very weird thing is that, if you modify `x_absmax = x.abs().amax(dim=-1, keepdim=True)` to `x_absmax = x.abs().amax(dim=-1, keepdim=True) * 0.4999` in the `get_scale` and `w4_quant` function in my script, **large diff occurs.**
> 
> I just tried it, and found this.
> 
> ```
> out1
> tensor([[-0.0216,  0.5732,  0.4524,  ..., -0.4541,  0.2771, -0.4653],
>         [-0.0375, -0.0845,  0.6333,  ...,  0.2949,  0.4790,  0.6089],
>         [-0.1888, -0.1674, -0.4871,  ...,  0.2385,  0.7886,  0.2507],
>         ...,
>         [-0.0891,  0.2343,  0.1569,  ...,  0.6831, -0.1757, -0.5054],
>         [-0.4187,  0.5977, -0.0225,  ...,  0.3330,  0.1484, -0.7993],
>         [-0.4817, -0.1526, -0.1072,  ...,  0.4109, -0.0023, -0.5581]],
>        device='cuda:0', dtype=torch.float16)
> out2
> tensor([[ 0.1066, -0.5508, -0.7207,  ...,  0.5381, -0.3894,  0.0118],
>         [ 0.0127, -0.4534, -0.3826,  ..., -0.1334, -0.0036, -0.2688],
>         [-0.1793, -0.0451,  0.4478,  ...,  0.4104,  0.6265,  0.5024],
>         ...,
>         [ 0.0429,  0.4233,  0.3572,  ..., -0.1161, -0.4062, -0.0836],
>         [ 0.2568,  0.0115,  0.0422,  ...,  0.2424,  0.0394,  0.2054],
>         [ 0.1392, -0.0248, -0.1234,  ..., -0.1874, -0.1741,  0.2067]],
>        device='cuda:0', dtype=torch.float16)
> ```

HandH1998 commented:
@brisker Thanks for your report! I have confirmed it is a bug in `pack` function. This is because that we missed `clamp` operation for per-channel quantization. So when the quant scale is calculated from the `absmax`, the diff comes out. Now I have fixed it [https://github.com/HandH1998/QQQ/commit/ae9531ce7a4d92080704df1e2b32f18a79600ae7](url). You can try your script again to verify it.

brisker commented:
@HandH1998 
after bug fixed, I have tried it, now is ok

HandH1998 commented:
@tlrmchlsmth Thanks for your review! I plan to resolve conflicts on Thursday. Then you can go ahead.

brisker commented:
@HandH1998 
Since many models are trained in bf16, is bf16 supported for this w4a8 PR?

HandH1998 commented:
@brisker BF16 is not supported in this PR. We don't have enough time to support it for now, though we would like to put it in our list. If you are interested in it, we welcome you submit a PR to our QQQ repo. This mainly includes changes about: [offline quant](https://github.com/HandH1998/QQQ/blob/main/QQQ/gptq/qlinear/qlinear_marlin.py) and [per-group dequant kernel](https://github.com/HandH1998/QQQ/blob/main/csrc/qqq_gemm.cu#L167). The former may be easy to achieve.

HandH1998 commented:
@tlrmchlsmth I have resolved conflicts.

brisker commented:
@HandH1998 

A few days ago, I said ok, but that is for w4a8-with-no-group.

Today I tried w4a8-gs128, now `assert(torch.allclose(out1, out2, atol=1e-3, rtol=1e-3))`  **is not ok**. ---this time, absmax or absmax*0.4999 **are all not ok.**

I think maybe there is another bug.

The new `get_scale `and `w4_quant `code is here:
```
def get_scale(x,group_size=-1):
    if group_size==-1:
        q_max = (2 ** (4-1)) - 1
        x_absmax = x.abs().amax(dim=-1, keepdim=True)*clip_ratio
        # if self.lwc:
        #     x_absmax = x_absmax * self.sigmoid(self.absbound_factor)

        scale = x_absmax / q_max
        return scale,None
    elif group_size==128:
        q_max = 2 ** (4) - 1
        zero_point = (q_max +1)/2
        reshaped_x = x.view(x.shape[0],int(x.shape[1]//group_size),group_size)
        xmax = reshaped_x.abs().amax(dim=-1, keepdim=True)*clip_ratio
        # if self.lwc:
        #     xmax = xmax * self.sigmoid(self.absbound_factor)
        xmin = -xmax
        scale = (xmax - xmin) / q_max
        # scale = scale.clamp(min=CLIPMIN, max=CLIPMAX)

        quant_value = torch.clamp(torch.round(reshaped_x / scale) + zero_point, 0, q_max)
        x_dequant =  scale * (quant_value - zero_point)
        x_dequant = x_dequant.view_as(x)


        ############ second step ##########
        second_step_q_max = 127  # int8
        absmax = x_dequant.abs().amax(dim=1, keepdim=True)
        scale_1 = absmax / second_step_q_max
        # import pdb;pdb.set_trace()
        # scale_1 = scale_1.clamp(min=CLIPMIN, max=CLIPMAX)
        return scale[:,:,0],scale_1

    else:
        raise RuntimeError
```

```
def w4_quant(x,group_size=-1):
    if group_size==-1:
        q_max = (2 ** (4-1)) - 1
        x_absmax = x.abs().amax(dim=-1, keepdim=True)*clip_ratio
        # x_absmax = x_absmax * 0.4999
        # if self.lwc:
        #     x_absmax = x_absmax * self.sigmoid(self.absbound_factor)
        scale = x_absmax / q_max
        # scale = scale.clamp(min=CLIPMIN, max=CLIPMAX)

        quant_value = torch.clamp(torch.round(x / scale), -q_max, q_max)

        x_dequant = quant_value * scale
        return x_dequant
    elif group_size==128:
        q_max = 2 ** (4) - 1
        zero_point = (q_max +1)/2
        reshaped_x = x.view(x.shape[0],int(x.shape[1]//group_size),group_size)
        # xmax = reshaped_x.amax(dim=-1, keepdim=True)
        # xmin = reshaped_x.amin(dim=-1,keepdim=True)  # bad direct w4a8-g128 acc, even worse than group=-1

        xmax = reshaped_x.abs().amax(dim=-1, keepdim=True)*clip_ratio
        # if self.lwc:
        #     xmax = xmax * self.sigmoid(self.absbound_factor)
        xmin = -xmax

        scale = (xmax - xmin) / q_max
        # scale = scale.clamp(min=CLIPMIN, max=CLIPMAX)

        quant_value = torch.clamp(torch.round(reshaped_x / scale) + zero_point, 0, q_max)
        x_dequant =  scale * (quant_value - zero_point)
        x_dequant = x_dequant.view_as(x)

        ###########


        ############ second step ##########
        second_step_q_max = 127  # int8
        absmax = x_dequant.abs().amax(dim=1, keepdim=True)
        scale_1 = absmax / second_step_q_max
        # scale_1 = scale_1.clamp(min=CLIPMIN, max=CLIPMAX)

        quant_value_1 = torch.clamp(torch.round(x_dequant / scale_1), -second_step_q_max, second_step_q_max)

        final_x_dequant = quant_value_1 * scale_1

        return final_x_dequant

    else:
        raise RuntimeError
```

> @HandH1998 @zhyncs I got a w4a8 llama2-7b-model with **wiki-ppl=5.7** when tested using fake-quant, but when I convert the weight into qqq-w4a8 format and test it again using the cuda-kernel, **the wiki-ppl becomes 10000.** So I run the following code to test the consistency between fake-quant and qqq-w4a8-kernel, I find there seems to be some difference which **can not be neglected.**
> 
> (I'm sure the conversion is right，because before tuning，the two ppl can be close，but all reasonable(7.5 and 7.53.) The `QuantLinear `class is extracted from [here](https://github.com/HandH1998/QQQ/blob/main/QQQ/gptq/qlinear/qlinear_marlin.py#L46) and `dynamic_quant` function is extracted from [here](https://github.com/HandH1998/QQQ/blob/main/QQQ/gptq/qlinear/qlinear_marlin.py#L246)
> 
> Any advice on this? I think this may be a potential bug.
> 
> **I think this issue is important because the carefully tuned accuracy is on the fake-quant domain, but there is gap between fake-quant and cuda-kernel, which means carefully tuned accuracy can not be real accuracy. We can not tune w4a8-accuracy on cuda-kernel-domain, right?**
> 
> ```
> from copy import deepcopy
> import torch
> import torch.nn as nn
> import sys 
> import QuantLinear
> 
> def dynamic_quant(x: torch.Tensor):
>     quant_scale = x.abs().max(dim=-1, keepdim=True)[0].div(127.0).to(torch.float)
>     quant_x = (x / quant_scale).round().clamp(-128, 127).to(torch.int8)
>     dequant_x = quant_x*quant_scale
>     return quant_x, quant_scale,dequant_x
> 
> def get_scale(x):
>     q_max = (2 ** (4-1)) - 1
>     x_absmax = x.abs().amax(dim=-1, keepdim=True)
> 
>     scale = x_absmax / q_max
> 
>     return scale,None
> 
> 
> def w4_quant(x):
>     q_max = (2 ** (4-1)) - 1
>     x_absmax = x.abs().amax(dim=-1, keepdim=True)
> 
>     scale = x_absmax / q_max
> 
> 
>     quant_value = torch.clamp(torch.round(x / scale), -q_max, q_max)
> 
>     x_dequant = quant_value * scale
> 
>     return x_dequant
> ori_fc = nn.Linear(2048,4096,bias=False).cuda().half()
> 
> qqq_linear = QuantLinear(
>                 4,
>                 -1,
>                 2048,
>                 4096,
>                 False,
>                 weight_dtype=torch.float16
>             )
> qqq_linear.cuda()
> 
> input = torch.Tensor(10,2048).normal_().cuda().half()
> quant_input, quant_input_scale,dequant_input = dynamic_quant(input)
> 
> 
> scale,scale_extra = get_scale(ori_fc.weight.data)
> qqq_linear.pack(ori_fc, scale, scale_extra)
> ori_fc.weight.data = w4_quant(ori_fc.weight.data)
> 
> with torch.no_grad():
>     out1 = ori_fc(dequant_input.half())
> 
> with torch.no_grad():
>     out2 = qqq_linear.forward(input)
>   
> print((out1==out2).sum(),out1.numel())
> ```

tlrmchlsmth commented:
@brisker this wouldn't be a problem with any of the code in this PR, right, but rather an issue in https://github.com/HandH1998/QQQ?

tlrmchlsmth commented:
BTW, I just tried rerunning the failing jobs

mgoin commented:
Marked PR as ready and resolved merge conflict, let's see if it is green now

HandH1998 commented:
@tlrmchlsmth @mgoin The failed jobs is using `lm_eval` to evaluate our model `HandH1998/QQQ-Llama-3-8b-g128`. I used the command `bash .buildkite/lm-eval-harness/run-lm-eval-gsm-vllm-baseline.sh -m HandH1998/QQQ-Llama-3-8b-g128 -b 32 -l 250 -f 5 -t 1` to get the evaluation result of `.buildkite/lm-eval-harness/configs/Meta-Llama-3-8B-QQQ.yaml` on my host before. The result is 0.484 for `exact_match,strict-match` and 0.492 for `exact_match,flexible-extract`. It can not match with the CR result (0.448 for `exact_match,strict-match`). 
Today I changed my flash-attn engine from `xformers` to `vllm-flash-attn` and got a new result `0.424` for `exact_match,strict-match` and 0.428 for `exact_match,flexible-extract`. It seems that flash-attn engine has a non-negligible impact on the final evaluation result.  Anyway, the result of my host still don't align with the result of CR.
 I wonder if you guys can reproduce the CR result. Please share your results here. Thanks!

mgoin commented:
Hi @HandH1998 I think you would see a more stable result if you increased the number of samples. Generally we use -l 1000 instead of 250

HandH1998 committed:
support qqq(w4a8) for vllm

HandH1998 commented:
@tlrmchlsmth @mgoin Hi, I have solved the CR issue when using -l 1000 instead of 250. However, new CR issues came out. I  think the new issues are not relevant with our code. Could you please take a look?

mgoin reviewed (approved):
Thanks for letting us know. The PP test is a flaky test so I will attempt to land
